File: driver/php/ext/class/class_group.php
<?php
/*******************************************************************************
   Copyright (C) 2012-2018 SequoiaDB Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*******************************************************************************/

/** \file class_group.php
    \brief group class
 */

/**
 * SequoiaGroup Class. To get this Class object must be call SequoiaDB::getGroup.
 */
class SequoiaGroup
{
   /**
    * Judge whether the specified replica group is catalog.
    *
    * @return Returns the result.
    *
    * @retval boolean isCatalog
    *
    * Example:
    * @code
    * $isCata = $groupObj -> isCatalog() ;
    * if( $isCata == true ) {
    *    echo "This is the catalog group." ;
    * } else {
    *    echo "This is not the catalog group" ;
    * }
    * @endcode
   */
   public function isCatalog(){}

   /**
    * Get the specified replica group name.
    *
    * @return Returns the group name.
    *
    * @retval string &lt;group_name&gt;
    *
    * Example:
    * @code
    * $groupName = $groupObj -> getName() ;
    * $err = $db -> getLastErrorMsg() ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to get group name, error code: ".$err['errno'] ;
    *    return ;
    * }
    * echo "Group name is: ".$groupName ;
    * @endcode
   */
   public function getName(){}

   /**
    * Start and activate the specified replica group.
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> start() ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to start group, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function start(){}

   /**
    * Stop the specified replica group.
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> stop() ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to stop group, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function stop(){}

   /**
    * Re-elected in the replication group.
    *
    * @param $options an array or the string argument. The $options of reelection. Please reference <a href="http://doc.sequoiadb.com/cn/sequoiadb-cat_id-1432190873-edition_id-@SDB_SYMBOL_VERSION">here</a> for more detail.
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example: 60 seconds for new elections.
    * @code
    * $err = $groupObj -> reelect( array( 'Seconds' => 60 ) ) ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to call reelect, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function reelect( array|string $options = NULL ){}

   /**
    * Get the group detail.
    *
    * @return Returns the group detail, default return array.
    *
    * @retval array   detail
    * @retval string  detail
    *
    * Example:
    * @code
    * $detail = $groupObj -> getDetail() ;
    * $err = $db -> getLastErrorMsg() ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to call getDetail, error code: ".$err['errno'] ;
    *    return ;
    * }
    * var_dump( $detail ) ;
    * @endcode
   */
   public function getDetail(){}

   /**
    * Get the master node.
    *
    * @return Returns a new SequoiaNode object.
    *
    * @retval SequoiaNode Object
    *
    * Example:
    * @code
    * $nodeObj = $groupObj -> getMaster() ;
    * if( empty( $nodeObj ) ) {
    *    $err = $db -> getLastErrorMsg() ;
    *    echo "Failed to get the master node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function getMaster(){}

   /**
    * Get the slave node.
    *
    * @param $positionsArray an array argument. The array of node's position, the array elements can be 1-7.
    *
    * @return Returns a new SequoiaNode object.
    *
    * @retval SequoiaNode Object
    *
    * Example: Get one of slave node
    * @code
    * $nodeObj = $groupObj -> getSlave() ;
    * if( empty( $nodeObj ) ) {
    *    $err = $db -> getLastErrorMsg() ;
    *    echo "Failed to get the slave node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
    *
    * Example: Get a specified slave node
    * @code
    * $nodeObj = $groupObj -> getSlave( array( 1, 2, 3 ) ) ;
    * if( empty( $nodeObj ) ) {
    *    $err = $db -> getLastErrorMsg() ;
    *    echo "Failed to get the slave node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function getSlave( $positionsArray = void ){}

   /**
    * Get the node.
    *
    * @param $name	the string argument. The name of the node.
    *
    * @return Returns a new SequoiaNode object.
    *
    * @retval SequoiaNode Object
    *
    * Example:
    * @code
    * $nodeObj = $groupObj -> getNode( 'host1:11910' ) ;
    * if( empty( $nodeObj ) ) {
    *    $err = $db -> getLastErrorMsg() ;
    *    echo "Failed to get the node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function getNode( string $name ){}

   /**
    * Create node.
    *
    * @param $hostName	the string argument. The hostname for the catalog replica group.
    *
    * @param $serviceName	the string argument. The servicename for the catalog replica group.
    *
    * @param $databasePath	the string argument. The path for the catalog replica group.
    *
    * @param $configure	the string argument. The configurations for the catalog replica group.
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> createNode( 'host1', '11900', '/opt/sequoiadb/database/catalog/11900' ) ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to create node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function createNode( string $hostName, string $serviceName, string $databasePath, array|string $configure = NULL ){}

   /**
    * Remove node in a given replica group.
    *
    * @param $hostName	the string argument. The hostname for the catalog replica group.
    *
    * @param $serviceName	the string argument. The servicename for the catalog replica group.
    *
    * @param $configure	the string argument. The configurations for the catalog replica group.
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> removeNode( 'host1', '11900' ) ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to remove node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function removeNode( string $hostName, string $serviceName, array|string $configure = NULL ){}

   /**
    * Attach a node to the group.
    *
    * @param $hostName	the string argument. The hostname for the catalog replica group.
    *
    * @param $serviceName	the string argument. The servicename for the catalog replica group.
    *
    * @param $options	an array or the string argument. The options of attach. Can be the follow options:
    *                                                  @code
    *                                                  KeepData : Whether to keep the original data of the new node. This option has no default value.
    *                                                             User should specify its value explicitly.
    *                                                  @endcode
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> attachNode( 'host1', '11900', array( 'KeepData' => true ) ) ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to attach node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function attachNode( string $hostName, string $serviceName, array|string $options ){}

   /**
    * Detach a node from the group.
    *
    * @param $hostName	the string argument. The hostname for the catalog replica group.
    *
    * @param $serviceName	the string argument. The servicename for the catalog replica group.
    *
    * @param $options	an array or the string argument. The options of detach. Can be the follow options:
    *                                                  @code
    *                                                  KeepData : Whether to keep the original data of the detached node. This option has no default value.
    *                                                             User should specify its value explicitly.
    *                                                  Enforced : Whether to detach the node forcibly, default to be false.
    *                                                  @endcode
    *
    * @return Returns the result, default return array.
    *
    * @retval array   array( 'errno' => 0 )
    * @retval string  { "errno": 0 }
    *
    * Example:
    * @code
    * $err = $groupObj -> detachNode( 'host1', '11900', array( 'KeepData' => true ) ) ;
    * if( $err['errno'] != 0 ) {
    *    echo "Failed to detach node, error code: ".$err['errno'] ;
    *    return ;
    * }
    * @endcode
   */
   public function detachNode( string $hostName, string $serviceName, array|string $options ){}
}

File: driver/php/ext/class/class_group.c
/*******************************************************************************
   Copyright (C) 2012-2014 SequoiaDB Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*******************************************************************************/

#include "class_group.h"

extern zend_class_entry *pSequoiadbNode ;

extern INT32 connectionDesc ;
extern INT32 groupDesc ;
extern INT32 nodeDesc ;

PHP_METHOD( SequoiaGroup, __construct )
{
}

PHP_METHOD( SequoiaGroup, isCatalog )
{
   zval *pThisObj = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   RETVAL_BOOL( sdbIsReplicaGroupCatalog( group ) ) ;
}

PHP_METHOD( SequoiaGroup, getName )
{
   INT32 rc = SDB_OK ;
   CHAR *pGroupName   = NULL ;
   zval *pThisObj     = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   pGroupName = (CHAR *)emalloc( PHP_NAME_SIZE ) ;
   if( !pGroupName )
   {
      rc = SDB_OOM ;
      goto error ;
   }
   ossMemset( pGroupName, 0, PHP_NAME_SIZE ) ;
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbGetRGName( group, pGroupName, PHP_NAME_LEN ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_RETVAL_STRING( pGroupName, 0 ) ;
done:
   return ;
error:
   RETVAL_EMPTY_STRING() ;
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, start )
{
   INT32 rc = SDB_OK ;
   zval *pThisObj = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbStartReplicaGroup( group ) ;
   if( rc )
   {
      goto error ;
   }
done:
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, stop )
{
   INT32 rc = SDB_OK ;
   zval *pThisObj = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbStopReplicaGroup( group ) ;
   if( rc )
   {
      goto error ;
   }
done:
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, reelect )
{
   INT32 rc = SDB_OK ;
   zval *pOptions = NULL ;
   zval *pThisObj = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   bson options ;
   bson_init( &options ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "|z", &pOptions ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   rc = php_auto2Bson( pOptions, &options TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbReelect( group, &options ) ;
   if( rc )
   {
      goto error ;
   }
done:
   bson_destroy( &options ) ;
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, getNodeNum )
{
   INT32 rc = SDB_OK ;
   INT32 status     = 0 ;
   INT32 count      = 0 ;
   zval *pStatus    = NULL ;
   zval *pGroupName = NULL ;
   zval *pSequoiadb = NULL ;
   zval *pThisObj   = getThis() ;
   sdbConnectionHandle connection = SDB_INVALID_HANDLE ;
   sdbCursorHandle cursor         = SDB_INVALID_HANDLE ;
   bson_type iteType              = BSON_EOO ;
   bson condition ;
   bson record ;
   bson_iterator it ;
   bson_init( &condition ) ;
   bson_init( &record ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "|z", &pStatus ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   /* 没有实现status
   rc = php_zval2Int( pStatus, &status TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   */
   PHP_READ_VAR( pThisObj, "_SequoiaDB", pSequoiadb ) ;
   PHP_READ_HANDLE( pSequoiadb,
                    connection,
                    sdbConnectionHandle,
                    SDB_HANDLE_NAME,
                    connectionDesc ) ;
   PHP_READ_VAR( pThisObj, "_name", pGroupName ) ;
   rc = bson_append_string( &condition,
                            FIELD_NAME_GROUPNAME,
                            Z_STRVAL_P( pGroupName ) ) ;
   if( rc != BSON_OK )
   {
      rc = SDB_DRIVER_BSON_ERROR ;
      goto error ;
   }
   bson_finish( &condition ) ;
   rc = sdbGetList( connection,
                    SDB_LIST_GROUPS,
                    &condition,
                    NULL,
                    NULL,
                    &cursor ) ;
   if( rc )
   {
      goto error ;
   }
   rc = sdbNext( cursor, &record ) ;
   if( rc )
   {
      goto error ;
   }
   iteType = bson_find( &it, &record, FIELD_NAME_GROUP ) ;
   if( iteType == BSON_ARRAY )
   {
      while( bson_iterator_more( &it ) )
      {
         ++count ;
      }
   }
   RETVAL_LONG( count ) ;
done:
   if( cursor != SDB_INVALID_HANDLE )
   {
      sdbCloseCursor( cursor ) ;
      sdbReleaseCursor( cursor ) ;
   }
   if( connection != SDB_INVALID_HANDLE )
   {
      sdbReleaseConnection( connection ) ;
   }
   bson_destroy( &condition ) ;
   bson_destroy( &record ) ;
   return ;
error:
   RETVAL_LONG( -1 ) ;
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, getDetail )
{
   INT32 rc = SDB_OK ;
   zval *pGroupName = NULL ;
   zval *pSequoiadb = NULL ;
   zval *pThisObj   = getThis() ;
   sdbConnectionHandle connection = SDB_INVALID_HANDLE ;
   sdbCursorHandle cursor         = SDB_INVALID_HANDLE ;
   bson condition ;
   bson record ;
   bson_init( &condition ) ;
   bson_init( &record ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_VAR( pThisObj, "_SequoiaDB", pSequoiadb ) ;
   PHP_READ_HANDLE( pSequoiadb,
                    connection,
                    sdbConnectionHandle,
                    SDB_HANDLE_NAME,
                    connectionDesc ) ;
   PHP_READ_VAR( pThisObj, "_name", pGroupName ) ;
   rc = bson_append_string( &condition,
                            FIELD_NAME_GROUPNAME,
                            Z_STRVAL_P( pGroupName ) ) ;
   if( rc != BSON_OK )
   {
      rc = SDB_DRIVER_BSON_ERROR ;
      goto error ;
   }
   bson_finish( &condition ) ;
   rc = sdbGetList( connection,
                    SDB_LIST_GROUPS,
                    &condition,
                    NULL,
                    NULL,
                    &cursor ) ;
   if( rc )
   {
      goto error ;
   }
   rc = sdbNext( cursor, &record ) ;
   if( rc )
   {
      goto error ;
   }
done:
   if( cursor != SDB_INVALID_HANDLE )
   {
      sdbCloseCursor( cursor ) ;
      sdbReleaseCursor( cursor ) ;
   }
   PHP_RETURN_AUTO_RECORD( FALSE,
                           pThisObj,
                           (rc == SDB_OK ? FALSE : TRUE),
                           record ) ;
   bson_destroy( &condition ) ;
   bson_destroy( &record ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, getMaster )
{
   INT32 rc = SDB_OK ;
   zval *pThisObj              = getThis() ;
   const CHAR *pNodeName       = NULL ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   sdbNodeHandle node          = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbGetNodeMaster( group, &node ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_BUILD_CLASS( FALSE,
                    pThisObj,
                    pSequoiadbNode,
                    node,
                    nodeDesc ) ;
   rc = sdbGetNodeAddr( node, NULL, NULL, &pNodeName, NULL ) ;
   if( rc == SDB_OK )
   {
      PHP_SAVE_VAR_STRING( return_value, "_name", pNodeName ) ;
   }
   rc = SDB_OK ;
done:
   return ;
error:
   RETVAL_NULL() ;
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, getSlave )
{
   INT32 rc = SDB_OK ;
   INT32 i  = 0 ;
   INT32 positionsCount        = 0 ;
   INT32 *pPositionsArray      = NULL ;
   zval *pThisObj              = getThis() ;
   zval *pPositions            = NULL ;
   const CHAR *pNodeName       = NULL ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   sdbNodeHandle node          = SDB_INVALID_HANDLE ;

   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;

   if ( PHP_GET_PARAMETERS( "|z", &pPositions ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }

   PHP_READ_HANDLE( pThisObj, group, sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME, groupDesc ) ;

   if ( pPositions )
   {
      rc = php_assocArray2IntArray( pPositions, &pPositionsArray,
                                    &positionsCount TSRMLS_CC ) ;
      if( rc )
      {
         goto error ;
      }
   }

   if( positionsCount > 0 )
   {
      rc = sdbGetNodeSlave1( group, pPositionsArray, positionsCount, &node ) ;
      if ( rc )
      {
         goto error ;
      }
   }
   else
   {
      rc = sdbGetNodeSlave( group, &node ) ;
      if( rc )
      {
         goto error ;
      }
   }

   PHP_BUILD_CLASS( FALSE,
                    pThisObj,
                    pSequoiadbNode,
                    node,
                    nodeDesc ) ;

   rc = sdbGetNodeAddr( node, NULL, NULL, &pNodeName, NULL ) ;
   if( rc == SDB_OK )
   {
      PHP_SAVE_VAR_STRING( return_value, "_name", pNodeName ) ;
   }

   rc = SDB_OK ;

done:
   if( pPositionsArray )
   {
      efree( pPositionsArray ) ;
   }
   return ;
error:
   RETVAL_NULL() ;
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, getNode )
{
   INT32 rc = SDB_OK ;
   PHP_LONG nodeLen = 0 ;
   CHAR *pNodeName = NULL ;
   zval *pThisObj  = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   sdbNodeHandle node          = SDB_INVALID_HANDLE ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "s", &pNodeName, &nodeLen ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbGetNodeByName( group, pNodeName, &node ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_BUILD_CLASS( FALSE,
                    pThisObj,
                    pSequoiadbNode,
                    node,
                    nodeDesc ) ;
   PHP_SAVE_VAR_STRING( return_value, "_name", pNodeName ) ;
done:
   return ;
error:
   RETVAL_NULL() ;
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, createNode )
{
   INT32 rc = SDB_OK ;
   PHP_LONG hostLen    = 0 ;
   PHP_LONG serviceLen = 0 ;
   PHP_LONG pathLen    = 0 ;
   CHAR *pHostName    = NULL ;
   CHAR *pServiceName = NULL ;
   CHAR *pPath        = NULL ;
   zval *pConfigure   = NULL ;
   zval *pThisObj     = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   bson configure ;
   bson_init( &configure ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "sss|z",
                           &pHostName,
                           &hostLen,
                           &pServiceName,
                           &serviceLen,
                           &pPath,
                           &pathLen,
                           &pConfigure ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   rc = php_auto2Bson( pConfigure, &configure TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbCreateNode( group,
                       pHostName,
                       pServiceName,
                       pPath,
                       &configure ) ;
   if( rc )
   {
      goto error ;
   }
done:
   bson_destroy( &configure ) ;
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, removeNode )
{
   INT32 rc = SDB_OK ;
   PHP_LONG hostLen    = 0 ;
   PHP_LONG serviceLen = 0 ;
   CHAR *pHostName    = NULL ;
   CHAR *pServiceName = NULL ;
   zval *pConfigure   = NULL ;
   zval *pThisObj     = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   bson configure ;
   bson_init( &configure ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "ss|z",
                           &pHostName,
                           &hostLen,
                           &pServiceName,
                           &serviceLen,
                           &pConfigure ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   rc = php_auto2Bson( pConfigure, &configure TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbRemoveNode( group,
                       pHostName,
                       pServiceName,
                       &configure ) ;
   if( rc )
   {
      goto error ;
   }
done:
   bson_destroy( &configure ) ;
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, attachNode )
{
   INT32 rc = SDB_OK ;
   PHP_LONG hostLen    = 0 ;
   PHP_LONG serviceLen = 0 ;
   CHAR *pHostName    = NULL ;
   CHAR *pServiceName = NULL ;
   zval *pConfigure   = NULL ;
   zval *pThisObj     = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   bson configure ;
   bson_init( &configure ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "ss|z",
                           &pHostName,
                           &hostLen,
                           &pServiceName,
                           &serviceLen,
                           &pConfigure ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   rc = php_auto2Bson( pConfigure, &configure TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbAttachNode( group,
                       pHostName,
                       pServiceName,
                       &configure ) ;
   if( rc )
   {
      goto error ;
   }
done:
   bson_destroy( &configure ) ;
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaGroup, detachNode )
{
   INT32 rc = SDB_OK ;
   PHP_LONG hostLen    = 0 ;
   PHP_LONG serviceLen = 0 ;
   CHAR *pHostName    = NULL ;
   CHAR *pServiceName = NULL ;
   zval *pConfigure   = NULL ;
   zval *pThisObj     = getThis() ;
   sdbReplicaGroupHandle group = SDB_INVALID_HANDLE ;
   bson configure ;
   bson_init( &configure ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   if( PHP_GET_PARAMETERS( "ss|z",
                           &pHostName,
                           &hostLen,
                           &pServiceName,
                           &serviceLen,
                           &pConfigure ) == FAILURE )
   {
      rc = SDB_INVALIDARG ;
      goto error ;
   }
   rc = php_auto2Bson( pConfigure, &configure TSRMLS_CC ) ;
   if( rc )
   {
      goto error ;
   }
   PHP_READ_HANDLE( pThisObj,
                    group,
                    sdbReplicaGroupHandle,
                    SDB_GROUP_HANDLE_NAME,
                    groupDesc ) ;
   rc = sdbDetachNode( group,
                       pHostName,
                       pServiceName,
                       &configure ) ;
   if( rc )
   {
      goto error ;
   }
done:
   bson_destroy( &configure ) ;
   PHP_RETURN_AUTO_ERROR( FALSE, pThisObj, rc ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}